R/source.R

```diff
@@ -215,6 +207,49 @@ rust_function <- function(code, env = parent.frame(), ...) {
   rust_source(code = code, env = env, ...)
 }
 
+invoke_cargo <- function(
+  toolchain, specific_target,
+  dir, profile, stdout, stderr,
+  use_rtools) {
+  # Append rtools path to the end of PATH on Windows
+  if (
+    use_rtools &&
+    .Platform$OS.type == "windows" &&
+    nzchar(Sys.getenv("RTOOLS40_HOME"))
+  ) {
+    env_path <- Sys.getenv("PATH")
+    r_tools_path <-
+      normalizePath(
+        file.path(
+          Sys.getenv("RTOOLS40_HOME"), # {rextendr} targets R >= 4.0
+          paste0("mingw", ifelse(R.version$arch == "i386", "32", "64")),
+          "bin"
+        )
+      )
+    Sys.setenv(PATH = paste(env_path, r_tools_path, sep = .Platform$path.sep))
+    # This retores PATH when function returns, i.e. after cargo finishes.
+    on.exit(Sys.setenv(PATH = env_path))
```

Comment:
Better to set the exit handler before you modify the `PATH` variable, instead of after.